### PR TITLE
fix(material/datepicker): update active date on focus

### LINKED ISSUE #23483: bug(datepicker): navigating with ChromeVox keyboard shortcuts does not update selected date
#### Reproduction

Steps to reproduce:
1. Open the [Datepicker Examples](https://material.angular.io/components/datepicker/examples)
2. Go to the "Datepicker action buttons" demo
3.  Activate "Open calendar" button
4. Linear navigate forward (Launcher + Right Arrow in ChromeOS)
5. press space
 
 
#### Expected Behavior
It behaves the same as pressing the right arrow key. It highlights the next day and selects the next day on closing.

#### Actual Behavior
Screenreader announces the next day, but does not select that day when closing

#### Environment
OS| Browser | Screen reader
ChromeOS | Google Chrome 91.0.4472.167 | ChromeVox

##### Additional Notes
This is copied from an [internal bug report](http://b/195600355) found during a11y testing.

### PR DESCRIPTION
**Pull request by zarend**
When a date cell on the calendar receives focus, set the active date
to that cell. This ensures that the active date matches the date with
browser focus.

Previously, we set the active date on keydown and click, but that was
problematic for screen readers. That's because many screen readers trigger
a focus event instead of a keydown event when using screen reader
specific navigation (VoiceOver, Chromevox, NVDA).

Fixes #23483

### CODE REVIEW

File: src/material/datepicker/month-view.ts
Comment (zarend) [inline src/material/datepicker/month-view.ts:255]: This is 90% identical to the ` _dateSelected` method above. This code could be more DRY. I'm open to suggestions on how to refactor it.

Here's what I had in mind. Create 2 reusable functions with shared code between `dateBecomesActive` and `_dateSelected` to reduce duplicated code.

```
// final variable names subject to bike shedding
_normalizeNumericDateToThisMonth(date: number, normalizeTo: D) {
     const year = this._dateAdapter.getYear(normalizeTo);
    const month = this._dateAdapter.getMonth(normalizeTo);
    return this._dateAdapter.createDate(year, month, date);
}

_didUserEventChangeActiveOrSelectedDate(event: MatCalendarUserEvent<number>, referenceDate: DateRange<D> | D | null) {
    let rangeStartDate: number | null;
    let rangeEndDate: number | null;
    
    if (referenceDate instanceof DateRange) {
      rangeStartDate = this._getDateInCurrentMonth(referenceDate.start);
      rangeEndDate = this._getDateInCurrentMonth(referenceDate.end);
    } else {
      rangeStartDate = rangeEndDate = this._getDateInCurrentMonth(referenceDate);
    }
    
    return rangeStartDate !== date || rangeEndDate !== date;
}
```

File: src/material/datepicker/multi-year-view.ts
Comment (zarend) [inline src/material/datepicker/multi-year-view.ts:221]: This is nearly identical to the `_yearSelected` method above. I'm open to suggestions on how to refactor this, or if we should refactor it to share code.

File: src/material/datepicker/year-view.ts
Comment (zarend) [inline src/material/datepicker/year-view.ts:201]: This is nearly identical to the `_monthSelected` method above. I'm open to suggestions on how to refactor this, or if we should refactor it to share code.

**Comment by zarend** [comment]
This is working and ready for review now. @jelbourn , is minor the correct target for this PR? It adds a public member to `MatCalendarBody`, but the change should be non-breaking.

File: src/material/datepicker/month-view.ts
Comment (jelbourn) [reply src/material/datepicker/month-view.ts:255]: I don't fully follow why these two things need to work the same way; in particular, the active date is always going to be a single date (whereas selection can be a range of dates). Shouldn't this be as direct as when a date-cell is focused, set that cell's date as the active date?

Comment (zarend) [reply src/material/datepicker/month-view.ts:255]: that's right, the active date can only be a single date. I was confused. That means I can delete the daterange code from `_dateBecomesActive`, which will clean the code up a bit.

**Comment by zarend** [comment]
Looks like the focus is lost when pressing Page Up in the month view. I'm looking into that...

File: src/material/datepicker/month-view.ts
Comment (zarend) [reply src/material/datepicker/month-view.ts:255]: I removed the unnecessary code.

**Comment by zarend** [comment]
> Looks like the focus is lost when pressing Page Up in the month view. I'm looking into that...

Fixed the losing focus problem.

**Comment by zarend** [comment]
Everything works when I test manually, so it appears to be a problem with the tests.

File: src/material/datepicker/calendar-body.ts
Comment (jelbourn) [inline src/material/datepicker/calendar-body.ts:160]: Avoid using `any` where possible. I believe this would be `FocusEvent`

File: src/material/datepicker/month-view.ts
Comment (jelbourn) [inline src/material/datepicker/month-view.ts:259]: Is this all still necessary? `getYear`, `getMonth`, etc. all take a date `D`, which you're then composing into... another date `D`. I think that the other function you were referencing did something like this to deal with ranges. You should be able to just do
```ts
    if (!this._dateAdapter.sameDate(event.value, this._activeDate)) {
      this.activeDateChange.emit(activeDate);
    }
```

We should also confirm that the `markForCheck` is actually necessary here

**Comment by zarend** [comment]
@crisbeto https://github.com/angular/components/pull/24172/commits/c0f8ecf2626815d77b4421042ac9cf4db49fb4de
^ calling `detectChanges` seems to fix the failing tests.

After debugging this with `console.log` statements, I determined that the calendar body calls `.focus` on the active cell before receiving an updated value for `activeCell`. This causes it to focus on the old active cell, rather than the one set in the keydown handler.

Ensuring that change detection runs appears to be the fix here, but I'm curious why this wan't a problem until now.

File: src/material/datepicker/month-view.ts
Comment (zarend) [reply src/material/datepicker/month-view.ts:259]: yes, constructing a date `D` from the event is really needed. That's because the event.value is a number that corresponds to the day of the month that becomes active. We turn that into a date `D` relative to the month and year of the current date.

The code from `_dateSelected` that references `rangeStartDate` and `rangeEndDate` is not needed here because activeDate doesn't needed to deal with ranges, but we really do need to convert the `MatCalendarUserEvent<number>` into a date `D`.

Comment (zarend) [reply src/material/datepicker/month-view.ts:259]: I don't think the `markForCheck` is necessary here because this method doesn't change any instance values on the component, so change detection doesn't seem relevant.

**Comment by zarend** [comment]
Note that this is not going to fix the problem with VoiceOver navigation not updating the focused date until #24171 lands. That's because VoiceOver does not seem to trigger the `focus` event when navigating to a `td` element, but it does when navigating to a `button element`.

This PR and #24171 are safe to merge in either order, but #23483 will not be fixed until both PR's land.

File: src/material/datepicker/month-view.ts
Comment (jelbourn) [reply src/material/datepicker/month-view.ts:259]: Since the month and year are the same, though, can't you just compare the dates, then?

File: src/material/datepicker/month-view.ts
Comment (jelbourn) [inline src/material/datepicker/month-view.ts:344]: Why is this necessary? This would invoke an entire extra change detection, which we generally want to avoid.

File: src/material/datepicker/month-view.ts
Comment (zarend) [reply src/material/datepicker/month-view.ts:259]: Yeah, that should work since we're not actually assigning to `activeDate`. I was confused between this method and `_handleCalendarBodyKeydown`.

Comment (zarend) [reply src/material/datepicker/month-view.ts:259]: We can do that for the check if the active date changed, but we still need a date `D` to pass to `activeDateChange`.

Comment (jelbourn) [reply src/material/datepicker/month-view.ts:259]: Ack, that makes sense

File: src/material/datepicker/month-view.ts
Comment (crisbeto) [reply src/material/datepicker/month-view.ts:344]: Yeah, this seems a little iffy. I'm guessing that it might be flushing this subscription: https://github.com/angular/components/blob/master/src/material/datepicker/calendar-body.ts#L201

Comment (zarend) [reply src/material/datepicker/month-view.ts:344]: After debugging this with console.log statements, I determined that the calendar body calls .focus on the active cell before receiving an updated value for activeCell. This causes it to focus on the old active cell, rather than the one set in the keydown handler.

Ensuring that change detection runs appears to be the fix here, but I'm curious why this wan't a problem until now.

**Review by josephperrott — APPROVED** [review]
LGTM

File: src/material/datepicker/month-view.ts
Comment (jelbourn) [reply src/material/datepicker/month-view.ts]: If it works in real life but not in tests, it most likely means the _test_ needs a `detectChanges` call and not the source code (which is very normal since in the test nothing is causing change dtections to happen autmatically)

Comment (zarend) [reply src/material/datepicker/month-view.ts]: I just pushed a change to only call `detectChanges` if the event is not trusted. This is only needed for the tests because fake keyboard events do not trigger change detection. I'm generally not a fan of having code that's only intended to run in the tests, and I'm curious how we usually handle this situation.

Calling detectChanges in the test code doesn't work because by that time the calendar body will have already triggered the `activeDateChange` event with a stale value. This wasn't a problem until now, because before focusing on a stale date wouldn't trigger the `activeDateChange` event to update the active date. Calling detectChanges in the test code would case the calendar body to focus again but on a updated value.

Here's a rough timeline of what happens in the unit tests if we don't call `detectChanges` when handling the `keydown` event.
0. (active Date is Jan 12 2021)
1. trigger a fake keyboard event with the right arrow key
2. MatMonthView sets `this.activeDate` to Jan 13 2021
3. MatMonthView calls `_focusActiveDate`
4. MatCalendarBody calls .focus on Jan 12, 2021 date then triggers an `activeDateChange` Event with Jan 12, 2021
5. test code calls `detectChanges`
6. month-view receives sets the active date to Jan 12, 2021
7. (active date should be Jan 13, 2021).

Comment (wagnermaciel) [reply src/material/datepicker/month-view.ts]: Is there a supported way to do this programmatically without dispatching fake events? If there isn't, teams will probably try dispatching keydown events manually to get this desired behavior. So if the support for this gets baked into the datepicker it will likely be depended on

Comment (jelbourn) [reply src/material/datepicker/month-view.ts]: > I'm generally not a fan of having code that's only intended to run in the tests, and I'm curious how we usually handle this situation.

This isn't really the way to think about this situation. In a real application, Angular will trigger change detection when a user interacts with the page (clicks, keypresses, etc). However, that won't happen if we trigger a fake event in a test (since obviously you can't trigger a real event in a test). So calling `detectChanges` in the test is the thing that most closely approximates what happens when the real app is running. The library code should almost _never_ have unnecessary behaviors added to support tests (barring things like extra attrs we add for harnesses).

Comment (zarend) [reply src/material/datepicker/month-view.ts]: Okay, so I should think about if the behavior is different between the tests and a real application, rather than the exact code path.

> The library code should almost never have unnecessary behaviors added to support tests

Does the library code mean Angular Material (this repo) or the Angular framework?

The above diff is out-of-date, but I made it only call `detectChanges` if the event is not trusted, so that we don't impact the performance in a real application.

**Comment by zarend** [comment]
@jelbourn I've responded to the PR comments and this is ready for your eyes again 👀 . Could you confirm that `target: minor` is appropriate for this PR? It makes an API change, but it does it in a non-breaking way. (cc @crisbeto also for Datepicker)

File: src/material/datepicker/month-view.ts
Comment (jelbourn) [inline src/material/datepicker/month-view.ts:346]: This isn't really the right approach to this issue. The unit test should directly call `detectChanges`- library code (this code) shouldn't add additional behaviors explosively to make unit tests pass. Having the tests call `detectChanges` in tandem with emitting the fake event is the normal/standard way of handling this.

File: src/material/datepicker/year-view.ts
Comment (jelbourn) [inline src/material/datepicker/year-view.ts:218]: Would the `clampDate` method on `DateAdapter` be a simpler way of implementing this?

File: src/material/datepicker/calendar-body.ts
Comment (crisbeto) [inline src/material/datepicker/calendar-body.ts:126]: nit: might be more accurate to call this `activeDateChange` to align with the events on the other calendar views.

**Review by crisbeto — APPROVED** [review]
LGTM assuming the last couple of comments will be resolved.

File: src/material/datepicker/month-view.ts
Comment (zarend) [reply src/material/datepicker/month-view.ts:346]: Well, calling `detectChanges` in the unit test doesn't work in this situation because the microtask queue is flushing before change detection runs. the child component (`MatCalendarBody`) needs to received the updated data from `this.activeDate` before it can focus on the correct cell in this method https://github.com/angular/components/blob/194ec1b7aad7df4b3deed4820c983151d0bad978/src/material/datepicker/calendar-body.ts#L209.

Here's an explanation I wrote up earlier which seems to have gotten buried in the PR comments.

I just pushed a change to only call detectChanges if the event is not trusted. This is only needed for the tests because fake keyboard events do not trigger change detection. I'm generally not a fan of having code that's only intended to run in the tests, and I'm curious how we usually handle this situation.

Calling detectChanges in the test code doesn't work because by that time the calendar body will have already triggered the activeDateChange event with a stale value. This wasn't a problem until now, because before focusing on a stale date wouldn't trigger the activeDateChange event to update the active date. Calling detectChanges in the test code would case the calendar body to focus again but on a updated value.

Here's a rough timeline of what happens in the unit tests if we don't call detectChanges when handling the keydown event.
0. (active Date is Jan 12 2021)
1. trigger a fake keyboard event with the right arrow key
2. MatMonthView sets this.activeDate to Jan 13 2021
3. MatMonthView calls _focusActiveDate
4. MatCalendarBody calls .focus on Jan 12, 2021 date then triggers an activeDateChange Event with Jan 12, 2021
5. test code calls detectChanges
6. month-view receives sets the active date to Jan 12, 2021
7. (active date is Jan 12, but it should be Jan 13, 2021).

There is two approaches I can think of, but both required changing the library code (this code)
1. call detect changes in library code (this code) only if the event is not trusted
2. add a `timeOut` to `_focusActiveCell` so that the microtask queue does not get flushed until we manually flush it in the test code.

WDYT?

Comment (wagnermaciel) [reply src/material/datepicker/month-view.ts:346]: If it is too difficult to handle testing these in unit tests, I would suggest just test these cases in our e2e tests with real keyboard events

**Comment by zarend** [comment]
Talked this over with @andrewseguin , This solution is problematic for testing because change detections needs to run before the microtask queue is flushed. We also looked into making `MatCalendarBody.activeCell` have a setter that calls `_focusActiveDate`, but that seemed to cause a regression with stealing the user's focus (see #11049).

Closing this PR,  I'm looking into another approach, where we would be able to run change detection in the tests. Will send out a new PR when I have it working.

**Comment by zarend** [comment]
FYI this is closed in favor of #24279 24279

**Pull request closed without merging**
